src/lib/code-sample-editor-preview.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  customElement,
+  LitElement,
+  html,
+  css,
+  property,
+  query,
+} from 'lit-element';
+import { ifDefined } from 'lit-html/directives/if-defined.js';
+import '@material/mwc-icon-button';
+import '@material/mwc-textfield';
+
+@customElement('code-sample-editor-preview')
+export class CodeSampleEditorPreviewElement extends LitElement {
+  static styles = css`
+    * {
+      box-sizing: border-box;
+    }
+
+    :host {
+      display: flex;
+      flex-direction: column;
+      width: 400px;
+    }
+
+    #toolbar {
+      display: flex;
+      flex-direction: row;
+      align-items: center;
+      flex: 0 0 36px;
+      color: #444;
+      background: #f0f0f0;
+      border-bottom: 1px solid #ddd;
+      padding: 0 4px;
+      --mdc-typography-button-font-size: 0.75rem;
+      font-size: var(--mdc-typography-button-font-size, 0.75rem);
+    }
+
+    #toolbar mwc-icon-button {
+      cursor: default !important;
+      --mdc-icon-button-size: 32px;
+      --mdc-icon-size: 20px;
+    }
+
+    #location {
+      height: 24px;
+      line-height: 18px;
+      /* border-radius: 4px; */
+      padding: 4px;
+      /* border: solid 1px #ccc;
+      outline: none; */
+      font-family: var(
+        --mdc-typography-button-font-family,
+        var(--mdc-typography-font-family, Roboto, sans-serif)
+      );
+    }
+
+    iframe {
+      flex: 1;
+      width: 100%;
+      border: none;
+    }
+  `;
+
+  @property()
+  src?: string;
```

Comment:
I need undefined here to represent the uninitialized case and to work with `ifDefined` before. I much prefer using `undefined` as the uninitialized value for JS objects, when not emulating DOM APIs, because it's the default value for a property any way. This shouldn't cause any problems with debugging. What have you seen?

Reply:
often issues with components are startup states and I've often run into issues where debugging we don't know what part of the render cycle it was in because the default value was `undefined` rather than our initialization of `null`. IMO `ifDefined` needs a parent function / directive that can take a comparison basis. See textfield:

https://github.com/material-components/material-components-web-components/blob/master/packages/textfield/src/mwc-textfield-base.ts#L317

Also counterpoint, `iframe.src` is an empty string if undefined and so is `input.value`

Reply:
Since defaults are set in the constructor, independent of any LitElement lifecycle, I don't think there should be a question of when in the render cycle you are. undefined doesn't give you any different information than null, since the constructor has to run.